core/src/main/java/org/elasticsearch/action/main/MainResponse.java

```diff
@@ -114,4 +117,24 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.endObject();
         return builder;
     }
+
+    private final static ObjectParser<MainResponse, Void> PARSER = new ObjectParser<>(MainResponse.class.getName(),
+            () -> new MainResponse());
+
+    static {
+        PARSER.declareString((response, value) -> response.nodeName = value, new ParseField("name"));
+        PARSER.declareString((response, value) -> response.clusterName = new ClusterName(value), new ParseField("cluster_name"));
+        PARSER.declareString((response, value) -> response.clusterUuid = value, new ParseField("cluster_uuid"));
+        PARSER.declareString((response, value) -> {}, new ParseField("tagline"));
+        PARSER.declareObject((response, value) -> {
+            response.build = new Build((String) value.get("build_hash"), (String) value.get("build_date"),
+                    (boolean) value.get("build_snapshot"));
+            response.version = Version.fromString((String) value.get("number"));
+            response.available = true;
+        }, (parser, context) -> parser.map(), new ParseField("version"));
```

Comment:
I already tested this manually, there is also a unit test for how the flag works here already: https://github.com/elastic/elasticsearch/blob/f8c28711bee711df87bab6b391ef2fee8a45a6c9/core/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java#L525

+1 for discussion how and where to make this configurable in a follow up.

When we miss fields (that is any parts of information) that are needed for constructing an object in Version X.Y we should barf I think, this usually happens when constructing the output object (either via constructor arg checks or some other validation). That problem is not limited by using (Constructing)ObjectParser I think. If some information is not there we either can proceed with a dummy value or we need to throw an error.